mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,505 @@
+package scala.meta.internal.pc
+import java.{util => ju}
+
+import scala.annotation.switch
+import scala.collection.mutable.ListBuffer
+
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.internal.pc.SemanticTokens._
+import scala.meta.pc.VirtualFileParams
+import scala.meta.tokens._
+
+import org.eclipse.lsp4j.SemanticTokenModifiers
+import org.eclipse.lsp4j.SemanticTokenTypes
+
+/**
+ * Corresponds to tests.SemanticHighlightLspSuite
+ */
+final class SemanticTokenProvider(
+    protected val cp: MetalsGlobal, // compiler
+    val params: VirtualFileParams
+) {
+  val capableTypes = TokenTypes
+  val capableModifiers = TokenModifiers
+
+  // Alias for long notation
+  val getTypeId: Map[String, Int] = capableTypes.zipWithIndex.toMap
+  val getModifierId: Map[String, Int] = capableModifiers.zipWithIndex.toMap
+
+  // Initialize Tree
+  import cp._
+  val unit: RichCompilationUnit = cp.addCompilationUnit(
+    params.text(),
+    params.uri().toString(),
+    None
+  )
+  cp.typeCheck(unit) // initializing unit
+  val (root, source) = (unit.lastBody, unit.source)
+
+  def unitPos(offset: Int): Position = unit.position(offset)
+  val nodes: List[NodeInfo] = traverser
+    .traverse(List.empty[NodeInfo], root)
+    .sortBy(_.pos.start)
+
+  /**
+   * main method
+   */
+  def provide(): ju.List[Integer] = {
+
+    val buffer = ListBuffer.empty[Integer]
+
+    import scala.meta._
+    var cLine = Line(0, 0) // Current Line
+    var lastProvided = SingleLineToken(cLine, 0, None)
+
+    for (tk <- params.text().tokenize.toOption.getOrElse(Nil)) yield {
```

Comment:
Do we actually need the `.toOption.getOrElse(Nil)`? If `.tokenize` is empty won't this just yield `Nil` anyways?